pkg/controller/machineset/machineset_controller.go

```diff
@@ -253,6 +262,58 @@ func (c *Controller) handleErr(err error, key interface{}) {
 	c.queue.Forget(key)
 }
 
+func (c *Controller) getMachineSet(key string) (*clusteroperator.MachineSet, error) {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return nil, err
+	}
+	if len(namespace) == 0 || len(name) == 0 {
+		return nil, fmt.Errorf("invalid key %q: either namespace or name is missing", key)
+	}
+	return c.machineSetsLister.MachineSets(namespace).Get(name)
+}
+
+func (c *Controller) checkClusterRunning(syncFunc func(string) error) func(string) error {
+	return func(key string) error {
+		err := syncFunc(key)
+		if err != nil {
+			return err
+		}
+		machineSet, err := c.getMachineSet(key)
+		if err != nil {
+			return err
+		}
+		cluster, err := controller.ClusterForMachineSet(machineSet, c.clustersLister)
+		if err != nil {
+			return err
+		}
+		if cluster.Status.Running || !cluster.Status.Provisioned {
+			return nil
+		}
+		machineSets, err := controller.MachineSetsForCluster(cluster, c.machineSetsLister)
+		if err != nil {
+			return nil
+		}
+		if len(machineSets) != len(cluster.Spec.MachineSets) {
+			return nil
+		}
+		for _, ms := range machineSets {
+			if ms.Spec.NodeType == clusteroperator.NodeTypeMaster {
+				if !ms.Status.Provisioned || !ms.Status.Installed || !ms.Status.ComponentsInstalled || !ms.Status.NodeConfigInstalled {
```

Comment:
Should we be checking Accepted here as well for masters?

Reply:
masters are automatically joined to the cluster and don't need to run a separate accept step